File: domain/ticket.go
package domain

import "time"

type TicketStatus string

const (
	TicketStatusBacklog  TicketStatus = "backlog"
	TicketStatusProgress TicketStatus = "progress"
	TicketStatusDone     TicketStatus = "done"
)

func (s TicketStatus) Valid() bool {
	switch s {
	case TicketStatusBacklog, TicketStatusProgress, TicketStatusDone:
		return true
	default:
		return false
	}
}

type TicketWarning struct {
	Code    string `json:"code"`
	Message string `json:"message"`
}

type TicketSummary struct {
	ID     string       `json:"id"`
	Status TicketStatus `json:"status"`
	Title  string       `json:"title"`
	Repo   string       `json:"repo,omitempty"`
	// AdditionalRepos are extra repositories in scope beyond the primary Repo.
	// Order carries no meaning; keys must be unique and must not include Repo.
	AdditionalRepos []string        `json:"additional_repos"`
	Created         *time.Time      `json:"created,omitempty"`
	Updated         *time.Time      `json:"updated,omitempty"`
	References      []string        `json:"references"`
	HasConclusion   bool            `json:"has_conclusion"`
	Warnings        []TicketWarning `json:"warnings"`
}

// TicketOutcome distinguishes how a ticket conclusion ended: shipped commits,
// investigation/spike work that produced no commits, or an outright
// rejection.
type TicketOutcome string

const (
	TicketOutcomeCompleted   TicketOutcome = "completed"
	TicketOutcomeExploratory TicketOutcome = "exploratory"
	TicketOutcomeRejected    TicketOutcome = "rejected"
)

func (o TicketOutcome) Valid() bool {
	switch o {
	case TicketOutcomeCompleted, TicketOutcomeExploratory, TicketOutcomeRejected:
		return true
	default:
		return false
	}
}

type TicketConclusion struct {
	StartedAt       time.Time     `json:"started_at"`
	ConcludedAt     time.Time     `json:"concluded_at"`
	Agent           string        `json:"agent,omitempty"`
	Profile         string        `json:"profile,omitempty"`
	Outcome         TicketOutcome `json:"outcome"`
	RejectionReason string        `json:"rejection_reason,omitempty"`
	Commits         []CommitRef   `json:"commits"`
	Body            string        `json:"body"`
}

type Ticket struct {
	TicketSummary
	Body       string            `json:"body"`
	Conclusion *TicketConclusion `json:"conclusion"`
}

type TicketBoard struct {
	Backlog  []TicketSummary `json:"backlog"`
	Progress []TicketSummary `json:"progress"`
	Done     []TicketSummary `json:"done"`
}

type CreateTicketParams struct {
	Title           string
	Repo            string
	AdditionalRepos []string
	Body            string
	References      []string
	Now             time.Time
}

type EditTicketParams struct {
	OldString  string
	NewString  string
	ReplaceAll bool
	Now        time.Time
}

type UpdateTicketMetadataParams struct {
	Title           *string
	Repo            *string
	AdditionalRepos *[]string
	References      *[]string
	Now             time.Time
}

type MoveTicketParams struct {
	To TicketStatus
}

File: domain/ticket.ts
import type { CommitRef } from "./commit";

export type TicketStatus = "backlog" | "progress" | "done";

export interface TicketWarning {
  code: string;
  message: string;
}

export interface TicketSummary {
  id: string;
  status: TicketStatus;
  title: string;
  repo?: string;
  created?: string;
  updated?: string;
  references: string[];
  has_conclusion: boolean;
  warnings: TicketWarning[];
}

export type TicketOutcome = "completed" | "exploratory" | "rejected";

export interface TicketConclusion {
  started_at: string;
  concluded_at: string;
  agent?: string;
  profile?: string;
  outcome: TicketOutcome;
  rejection_reason?: string;
  commits: CommitRef[];
  body: string;
}

export interface Ticket extends TicketSummary {
  body: string;
  conclusion: TicketConclusion | null;
}

export interface TicketBoard {
  backlog: TicketSummary[];
  progress: TicketSummary[];
  done: TicketSummary[];
}

export interface CreateTicketParams {
  title: string;
  repo: string;
  body: string;
  references: string[];
  now: string;
}

export interface EditTicketParams {
  oldString: string;
  newString: string;
  replaceAll: boolean;
  now: string;
}

export interface UpdateTicketMetadataParams {
  title?: string;
  repo?: string;
  references?: string[];
  now: string;
}

export interface MoveTicketParams {
  to: TicketStatus;
}
